src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -157,19 +160,36 @@ private void streamToOutput() {
     }
 
     private void flushData(final Partition.PartitionKey pk, final SortedMap<Long, String> data) {
-        final String evt = EventStream.createStreamEvent(
-                pk.partition,
-                String.valueOf(offsets.get(pk).getSentOffset()),
-                new ArrayList<>(data.values()),
-                Optional.empty());
         try {
-            getOut().streamData(evt.getBytes(EventStream.UTF8));
+            final String offset = String.valueOf(offsets.get(pk).getSentOffset());
+            final String batch = serializeBatch(pk, offset, new ArrayList<>(data.values()));
+            getOut().streamData(batch.getBytes(EventStream.UTF8));
         } catch (final IOException e) {
             getLog().error("Failed to write data to output.", e);
             shutdownGracefully("Failed to write data to output");
         }
     }
 
+    private String serializeBatch(final Partition.PartitionKey partitionKey, final String offset,
+                                  final List<String> events) throws JsonProcessingException {
+
+        final String eventType = getContext().getEventTypesForTopics().get(partitionKey.topic);
+        final String token = getContext().getCursorTokenService().generateToken();
+        final SubscriptionCursor cursor = new SubscriptionCursor(partitionKey.partition, offset, eventType, token);
+        final String cursorSerialized = getContext().getObjectMapper().writeValueAsString(cursor);
+
+        final StringBuilder builder = new StringBuilder()
```

Comment:
Is it done with string builder for efficiency? Why not use json object mapper and serialize a Domain object?